[NCCL] Overwrite NCCL Stream in AllReduce

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43675 Test for NCCL Stream Overwrite
* #41356 [NCCL] Expose NCCL-specific collective options to Python
* #41276 [NCCL] Adding CUDA Stream Overwrite Options for Allgather, Reduce, ReduceScatter, and Broadcast
* **#41275 [NCCL] Overwrite NCCL Stream in AllReduce**

Allows user to pass a vector of CUDA stream objects through AllReduce Options, and use those stream instead of default streams to enqueue GPU operations.

Differential Revision: [D22242878](https://our.internmc.facebook.com/intern/diff/D22242878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22242878/)!

osalpekar committed:
[NCCL] Overwrite NCCL Stream in AllReduce

Allows user to pass a vector of CUDA stream objects through AllReduce Options, and use those stream instead of default streams to enqueue GPU operations.

Differential Revision: [D22242878](https://our.internmc.facebook.com/intern/diff/D22242878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22242878/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Overwrite NCCL Stream in AllReduce"


Allows user to pass a vector of CUDA stream objects through AllReduce Options, and use those stream instead of default streams to enqueue GPU operations.

Differential Revision: [D22242878](https://our.internmc.facebook.com/intern/diff/D22242878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22242878/)!

[ghstack-poisoned]

jiayisuse commented:
Seems we didn't update Python interface. Let's bind NCCLAllreduceOptions with python?

osalpekar commented:
> Seems we didn't update Python interface. Let's bind NCCLAllreduceOptions with python?

Added Python bindings in https://github.com/pytorch/pytorch/pull/41356

osalpekar committed:
Update on "[NCCL] Overwrite NCCL Stream in AllReduce"


Allows user to pass a vector of CUDA stream objects through AllReduce Options, and use those stream instead of default streams to enqueue GPU operations.

Differential Revision: [D22242878](https://our.internmc.facebook.com/intern/diff/D22242878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22242878/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Overwrite NCCL Stream in AllReduce"


Allows user to pass a vector of CUDA stream objects through AllReduce Options, and use those stream instead of default streams to enqueue GPU operations.

Differential Revision: [D22242878](https://our.internmc.facebook.com/intern/diff/D22242878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22242878/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Overwrite NCCL Stream in AllReduce"


Allows user to pass a vector of CUDA stream objects through AllReduce Options, and use those stream instead of default streams to enqueue GPU operations.

Differential Revision: [D22242878](https://our.internmc.facebook.com/intern/diff/D22242878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22242878/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Overwrite NCCL Stream in AllReduce"


Allows user to pass a vector of CUDA stream objects through AllReduce Options, and use those stream instead of default streams to enqueue GPU operations.

Differential Revision: [D22242878](https://our.internmc.facebook.com/intern/diff/D22242878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22242878/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Overwrite NCCL Stream in AllReduce"


Allows user to pass a vector of CUDA stream objects through AllReduce Options, and use those stream instead of default streams to enqueue GPU operations.

Differential Revision: [D22242878](https://our.internmc.facebook.com/intern/diff/D22242878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22242878/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Overwrite NCCL Stream in AllReduce"


Allows user to pass a vector of CUDA stream objects through AllReduce Options, and use those stream instead of default streams to enqueue GPU operations.

Differential Revision: [D22242878](https://our.internmc.facebook.com/intern/diff/D22242878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22242878/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -801,15 +801,25 @@ template <typename Fn, typename PreProcess, typename PostProcess>
+
+  // Check the size of the streams vector (this is necessary when we're using
+  // the user-supplied CUDA streams).
+  if (streams.size() != ncclStreams_[key].size()) {
```
Probably want to compare with `devices.size()` here.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -801,15 +801,25 @@ template <typename Fn, typename PreProcess, typename PostProcess>
 std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
     std::vector<at::Tensor>& inputs,
     std::vector<at::Tensor>& outputs,
+    std::vector<at::cuda::CUDAStream> cudaStreams,
```
pass as a const ref?

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -864,6 +874,21 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
+std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
+    std::vector<at::Tensor>& inputs,
+    std::vector<at::Tensor>& outputs,
+    std::vector<at::cuda::CUDAStream> cudaStreams,
```
const ref?

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -872,19 +897,33 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
   return collective(
       inputs,
       outputs,
+      std::vector<at::cuda::CUDAStream>(),
```
nit: I guess `{}` would work here?

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -872,19 +897,33 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
+    PreProcess pre,
+    PostProcess post) {
+  return collective(
+      inputs, outputs, std::vector<at::cuda::CUDAStream>(), fn, pre, post);
```
nit: `{}` here as well?

pritamdamania87 commented on torch/lib/c10d/Types.hpp:
```diff
@@ -25,8 +29,12 @@ struct BroadcastOptions {
+  virtual ~AllreduceOptions() = default;
   ReduceOp reduceOp = ReduceOp::SUM;
   std::chrono::milliseconds timeout = kUnsetTimeout;
+/* #ifdef USE_C10D_NCCL */
+/*   std::vector<at::cuda::CUDAStream> cudaStreams = {}; */
+/* #endif */
```
Why is the entire block commented out here?

pritamdamania87 reviewed (commented):
Could you add some details to the PR summary or link to an existing gh issue about the use case for this feature?

It looks like we only add these options in CPP for now, is there a plan to expose these options to python?
